keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,560 @@
+into the type defined below.  To support finer-grained configuration, instead
+of defining a single controller-wide transformation matrix, they could be embedded
+in ClusterQueue ResourceGroups (perhaps implicitly by instead embedding the
+transformation matrix in a ResourceFlavor).
+
+```go
+type ResourceTransformationStrategy string
+const Retain ResourceTransformationStrategy = "Retain"
+const Replace ResourceTransformationStrategy = "Replace"
+
+type ResourceTransformation struct {
+  // key is the output resource name and value is the quantity of the output resource 
+  // that corresponds to 1 unit of the input resource
+  ConversionBaseValues map[corev1.ResourceName]resource.Quantity `json:"conversionBaseValues"`
+
+  // should the input resource be replaced or retained
+  Strategy ResourceTransformationStrategy `json:"strategy"`
+}
+
+// The map key is the input resource name
+type ResourceTransformationMatrix map[corev1.ResourceName]ResourceTransformation
+
+```
+By using maps instead of arrays, we avoid needing
+to define the semantics of multiple entries having the same input or output
+resource name.  Using the input resource name as the key to the outer map
+avoids conflicting definitions of `Replace` for an input resource.
+
+As an example, the following list of 4-tuples
+
+* (`nvidia.com/mig-1g.5gb`, `kueue.x-k8s.io/accelerator-memory`, `5G`, `replace`)
+* (`nvidia.com/mig-2g.10gb`, `kueue.x-k8s.io/accelerator-memory`, `10G`, `replace`)
+* (`nvidia.com/mig-4g.20gb`, `kueue.x-k8s.io/accelerator-memory`, `20G`, `replace`)
+* (`nvidia.com/mig-1g.5gb`, `kueue.x-k8s.io/credits`, `10`, `replace`)
+* (`nvidia.com/mig-2g.10gb`, `kueue.x-k8s.io/credits`, `15`, `replace`)
+* (`nvidia.com/mig-4g.20gb`, `kueue.x-k8s.io/credits`, `30`, `replace`)
+* (`cpu`, `kueue.x-k8s.io/credits`, `1`, `retain`)
+
+would be represented in yaml when embedded in the Kueue Configuration object as
+```yaml
+resources:
+  excludedResourcePrefixes: []
+  transformations:
+    nvidia.com/mig-1g.5gb:
+      strategy: Replace
+      conversionBaseValues:
+        kueue.x-k8s.io/accelerator-memory: 5G
+        kueue.x-k8s.io/credits: 10
+    nvidia.com/mig-2g.10gb:
+      strategy: Replace
+      conversionBaseValues:
+        kueue.x-k8s.io/accelerator-memory: 10G
+        kueue.x-k8s.io/credits: 15
+    nvidia.com/mig-4g.20gb:
+      strategy: Replace
+      conversionBaseValues:
+        kueue.x-k8s.io/accelerator-memory: 20G
+        kueue.x-k8s.io/credits: 30
+    cpu:
+      strategy: Retain
+      conversionBaseValues:
+        kueue.x-k8s.io/credits: 1
+```
+
+### Granularity Options
+
+There are three natural options.
+
+1. Following how `excludedResourcePrefixes` works, define a single transformation matrix at the operator level.
+2. Allow per-ClusterQueue configuration by making the transformation matrix part of the ResourceGroup
+3. Define the transformation matrix as part of the ResourceFlavor definition.
```

Comment:
Even with option (3)  we may have redundancy to repeat the transformations across resource flavors (which may be representing spot vs on-demand vs reservation machines).  I would like to avoid the repeated transformations, as they will be hard to maintain by admins to edit consistently in multiple places.

Maybe another option is to use top-level API, like ResourceTransformationMatrix API. Then we could reference the objects from the RF or CQ instances. If done properly it would also not require Kueue restart to update.

OTOH, I want to keep it simple and grounded to the use cases, and the global configuration seems to majority of them. Even if we choose global to implement the global rules we may still add API support later. If `resourcetransformationMatrixName` is supplied, then use the indicated API object, otherwise fallback to the global.

WDYT @mwielgus @dgrove-oss ?

Reply:
It seems reasonable to me to start with option 1 as the simplest and then extend later with API support if needed.  I agree that adding  a reference `resourcetransformationMatrixName` as you suggested is the best option for avoiding duplication.

Reply:
I see, so I think we have two competitors:
I. global now, dedicated ResourceTransformationMatrix API if needed (as currently in KEP)
II. dedicated ResourceTransformationMatrix API from start with two variants - referenced from (a) CQ or  (b) RF

I guess doing (II)  is slightly more code, but the pros are: 

1. it allows us to go via Alpha API and so easily evolve the API
2. natural way of supporting multiple transformations  - no need to create 2 APIs (global and API) to do the same thing

It can be helpful to know how likely you see a need for multiple transformations in the future. @dgrove-oss @tardieu  do you have any insights here?

I'm now slightly leaning towards IIb (API object attached to RF). Any opinion @alculquicondor @mwielgus ?

Reply:
A single global matrix works well for the first set of user stories (MIG devices).  We are fairly confident we understand this use case and that it will be useful in practice. 

The second user story ("pseudo-currency" quotas) seems like it would work better if we could have different transformation matrixes for different resource flavors.  This user story is a bit more speculative, but it seems plausible and generated some interest when we discussed it in the batch wg call last month.

So I would also lean to IIb as the end goal.  I don't think the additional implementation complexity will be too bad, with the possible exception of the combination of flavor fungibility and per-flavor transformation matrices making it more complex to explain exactly why a ClusterQueue didn't have sufficient resources to admit a workload. 

We would need to add `resourcetransformationMatrixName` to ResourceFlavor.  Are there any issues in adding a field that is intended to hold the name of an alpha API instance ResourceFlavor

Reply:
> A single global matrix works well for the first set of user stories (MIG devices). We are fairly confident we understand this use case and that it will be useful in practice.

+1

> The second user story ("pseudo-currency" quotas) seems like it would work better if we could have different transformation matrixes for different resource flavors. This user story is a bit more speculative, but it seems plausible and generated some interest when we discussed it in the batch wg call last month.

Yes, it is speculative, but seems likely to happen as we move forward. 

>  I don't think the additional implementation complexity will be too bad, with the possible exception of the combination of flavor fungibility and per-flavor transformation matrices making it more complex to explain exactly why a ClusterQueue didn't have sufficient resources to admit a workload.

I know you have a working prototype using the global config, but it would be good to try the APIs and weigh the complexity added. Long term I think it is the cleanest approach.

> We would need to add resourcetransformationMatrixName to ResourceFlavor. Are there any issues in adding a field that is intended to hold the name of an alpha API instance to ResourceFlavor?

This is not an issue I think. We reference by name Alpha API objects from Beta, for example Cohort API is alpha and addressed from CQ. We also planned a similar approach for [Topology Aware Scheduling](https://github.com/kubernetes-sigs/kueue/tree/main/keps/2724-topology-aware-schedling#admin-facing-api) - see the TopologyName field referencing the Alpha Topology object.

I'm open to also consider alternative names for the API object, and the corresponding field, like: 
1. `resourcetransformationMatrixName` (and `ResourcetransformationMatrix` object) (as currently discussed)
2. `resourceMapperName` (and `ResourceMapper` object)
3. `resourceConfigName` (and `ResourceConfig` object) (if we want other configurations to also keep there)

Reply:
I like `resourceMapperName` and `ResourceMapper`.  It's shorter and also avoids thinking about whether the plural of matrix should be matrices or matrixes :).

In thinking about this some more, I think we might actually want `resourceMapperNames` on ResourceFlavor to be an array to allow maximal sharing of configuration.  Since ResourceFlavor currently doesn't have a Status, we'd defer the combining of the resourceMapperNames to the ClusterQueues that have ResourceGroups that use a ResourceFlavor with a non-empty resourceMapperNames. We could record the combined mapping in the ClusterQueue's status and if there were clashes when combining mappings we could make the ClusterQueue inactive until they were resolved and report the problem in the ClusterQueue's status. 

I'll update the KEP to propose IIb as the preferred alternative with a ResourceMapper API object and move the other configuration variants we discussed to the alternatives section.  Then go ahead and update the prototype to see what it would look like.

Reply:
> I like resourceMapperName and ResourceMapper. It's shorter and also avoids thinking about whether the plural of matrix should be matrices or matrixes :).

+1 

> I think we might actually want resourceMapperNames on ResourceFlavor to be an array to allow maximal sharing of configuration.

This sounds reasonable, but we need to balance the complication with usefulness, so some motivation use-case for that would be useful.

> Since ResourceFlavor currently doesn't have a Status

Would the story be different if it had status? Maybe we could consider adding it, but not sure about the process here since we usually add fields by pointers.

> I'll update the KEP to propose IIb as the preferred alternative with a ResourceMapper API object and move the other configuration variants we discussed to the alternatives section. Then go ahead and update the prototype to see what it would look like.

+1

Reply:
I like the idea of starting with I and leaving II.b for a possible follow up based on user feedback.

Reply:
That sounds good to me.
Actually, I mentioned a similar approach to the above.

Reply:
I've rewritten to propose a global configuration (I) and moved other options to alternatives.

Reply:
cool, thanks for the update